toolsrc/include/vcpkg/binarycaching.h

```diff
@@ -31,14 +35,27 @@ namespace vcpkg
         virtual ~IBinaryProvider() = default;
         /// Gives the BinaryProvider an opportunity to batch any downloading or server communication for executing
         /// `plan`.
-        virtual void prefetch(const VcpkgPaths& paths, const Dependencies::ActionPlan& plan) = 0;
+        void prefetch(const VcpkgPaths& paths, const Dependencies::ActionPlan& plan);
         /// Attempts to restore the package referenced by `action` into the packages directory.
-        virtual RestoreResult try_restore(const VcpkgPaths& paths, const Dependencies::InstallPlanAction& action) = 0;
+        virtual RestoreResult try_restore(const VcpkgPaths& paths, const Dependencies::InstallPlanAction& action);
         /// Called upon a successful build of `action`
-        virtual void push_success(const VcpkgPaths& paths, const Dependencies::InstallPlanAction& action) = 0;
+        virtual void push_success(const VcpkgPaths& paths, const Dependencies::InstallPlanAction& action);
+        /// Requests the result of `try_restore()` without actually downloading the package. Used by CI to determine
+        /// missing packages.
+        std::unordered_map<const Dependencies::InstallPlanAction*, RestoreResult> precheck(
+            const VcpkgPaths& paths, View<Dependencies::InstallPlanAction> actions);
+
+        friend struct MergeBinaryProviders;
+
+    protected:
+        virtual void prefetch(const VcpkgPaths& paths,
+                              const Dependencies::ActionPlan& plan,
+                              std::set<PackageSpec>* restored);
```

Comment:
```suggestion
                              std::set<PackageSpec>& restored);
```

This certainly isn't being used in a way that is pointerlike.